crates/storage/db/src/snapshot.rs

```diff
@@ -0,0 +1,83 @@
+//! reth's snapshot creation from database tables
+
+use crate::{
+    abstraction::cursor::DbCursorRO,
+    table::{Key, Table},
+    transaction::DbTx, RawKey, RawTable,
+};
+use reth_interfaces::RethResult;
+use reth_nippy_jar::NippyJar;
+use std::ops::RangeInclusive;
+
+/// Macro that generates snapshot creation functions that take an arbitratry number of [`Table`] and
+/// creates a [`NippyJar`] file out of their [`Table::Value`]. Each list of [`Table::Value`] from a
+/// table is a column of values.
+///
+/// Has membership filter set and compression dictionary support.
+macro_rules! generate_snapshot_func {
+    ($(($($tbl:ident),+)),+ $(,)? ) => {
+        $(
+            paste::item! {
+                /// Creates a snapshot from specified tables. Each table's `Value` iterator represents a column.
+                ///
+                /// **Ensure the range contains the same number of rows.**
+                ///
+                /// * `tx`: Database transaction.
+                /// * `range`: Data range for columns in tables.
+                /// * `keys`: List of keys (eg. `TxHash` or `BlockHash`) with length equal to `row_count` and ordered by future column insertion from `range`.
+                /// * `dict_compression_set`: Sets of column data for compression dictionaries. Max size is 2GB. Row count is independent.
+                /// * `row_count`: Total rows to add to `NippyJar`. Must match row count in `range`.
+                /// * `nippy_jar`: Snapshot object responsible for file generation.
+                #[allow(non_snake_case)]
+                pub fn [<create_snapshot$(_ $tbl)+>]<'tx,
+                    $($tbl: Table<Key=K>,)+
+                    K
+                >
+                (
+                    tx: &impl DbTx<'tx>,
+                    range: RangeInclusive<K>,
+                    dict_compression_set: Option<Vec<impl Iterator<Item = Vec<u8>>>>,
+                    keys: Option<Vec<Vec<u8>>>,
+                    row_count: usize,
+                    nippy_jar: &mut NippyJar
+                ) -> RethResult<()>
+                    where K: Key + Copy
+                {
+                    let range: RangeInclusive<RawKey<K>> = RawKey::new(*range.start())..=RawKey::new(*range.end());
+
+                    // Create PHF and Filter if required
+                    if let Some(keys) = keys {
+                        // This might get big but ph needs the values in memory.
```

Comment:
it concerns me a bit, what if we'll have multiple snapshots creating in background in parallel? Maybe we can use [CachedKeySet](https://docs.rs/ph/latest/ph/fmph/keyset/enum.CachedKeySet.html) to pass an iterator?

Reply:
That would require a few more traits to be implemented (Send + Sync + Clone) for the iterator, which would touch on the mdbx cursor and walker implementations.

I think it's doable, but it requires a bit more effort and complexity. For now, I made `prepare_index` to take an iterator, but it's still collecting it inside. Once we make those mentioned changes, we can change it accordingly.